lib/teleterm/services/clientcache/clientcache.go

```diff
@@ -0,0 +1,177 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package clientcache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+)
+
+// Cache stores remote clients keyed by cluster URI.
+// Safe for concurrent access.
+// Closes all clients and wipes the cache on Close.
+type Cache struct {
+	clusters.Resolver
+	mu  sync.Mutex
+	log *logrus.Entry
+	// clients keep mapping between cluster URI
+	// (both root and leaf) and proxy clients
+	clients map[uri.ResourceURI]*client.ProxyClient
+	// group prevents duplicate requests to create remote clients
+	// for a given cluster URI
+	group singleflight.Group
+}
+
+// Config describes the client cache configuration.
+type Config struct {
+	clusters.Resolver
+	Log *logrus.Entry
+}
+
+// New creates an instance of Cache.
+func New(c Config) *Cache {
+	return &Cache{
+		log:      c.Log,
+		clients:  make(map[uri.ResourceURI]*client.ProxyClient),
+		Resolver: c.Resolver,
+	}
+}
+
+// Get returns a proxy client from the cache if there is one,
+// otherwise it dials the remote server.
+func (c *Cache) Get(ctx context.Context, clusterURI uri.ResourceURI) (*client.ProxyClient, error) {
```

Comment:
It might be worth mentioning that clients returned by `Get` should not get closed by the caller.